Add compile-fail test in rustdoc

GuillaumeGomez opened the pull request:
r? @brson 
cc @alexcrichton 

I still need to add error code explanation test with this, but I can't figure out a way to generate the `.md` files in order to test example source codes.

Will fix #27328.

steveklabnik commented:
This feature will be real awesome.

alexcrichton commented:
I personally don't think that this ends up being too useful of a feature as-is in practice. Knowing that an example doesn't compile provides no guarantee about preventing something from accidentally regressing in the future because it's far more likely to break for other reasons such as API change.

Along those lines the only "really useful" form of compile-fail tests in my opinion are the ones we have which check error messages. Those are quite flaky, however, as the order, number, and placement of errors changes over time in the compiler. 

Overall I don't think that flagging an example as purely "this should not compile" is a useful feature in the long run, and the more useful "this should not compile with this message" is not something we want to provide because it is not stable.

rphmeier commented on src/librustdoc/test.rs:
```diff
@@ -247,10 +248,18 @@ fn runtest(test: &str, cratename: &str, cfgs: Vec<String>, libs: SearchPaths,
+        driver::compile_input(sess, &cstore, cfg, &input, &out, &None, None, control);
+    }).join() {
+        Ok(_) if compile_fail           => panic!("couldn't compile the test"),
+        Err(_) if compile_fail == false => panic!("test compiled while it wasn't supposed to"),
```
Doesn't Ok here mean that the code does compile (insofar as compilation doesn't panic)? The match here seems to indicate the opposite.

GuillaumeGomez replied:
@rphmeier: Indeed, I switched error messages. Thanks for notifying me.

GuillaumeGomez commented:
@alexcrichton: The point here isn't to prevent regression but to check if examples are still up to date. For the rest, I don't think I'm the best to answer.

brson commented:
> Overall I don't think that flagging an example as purely "this should not compile" is a useful feature in the long run, and the more useful "this should not compile with this message" is not something we want to provide because it is not stable.

This was intended to be the first step. Testing that examples don't compile is better than not testing them, and we can improve later with more precise checking of errors.

Compile-fail tests are an important type of test, and none of our documentation examples that don't compile are tested. We are planning on using this to test the compiler error index, which is filled with such examples.

Given that compile-fail doc tests are important, it is clear to me that the tool we should use to test them is rustdoc, the tool we do all our other documentation testing with.

Do you have other suggestions about how we might test documentation examples that don't compile?

brson commented:
For the basic case of testing that examples fail to compile, without checking the specific errors, this patch lgtm, barring @alexcrichton's concerns.

alexcrichton commented:
I guess what I was getting at here was two-fold:

1. As-is I don't necessarily think that this is too useful to have in the long run.
2. With the logical extension of testing for precise messages, I don't think we want to commit to that because of stability concerns.

I agree, though, that testing the error index is something we should definitely do, so I'd be happy with ensuring that this is feature gated. We can always stomach changing our own error messages, but it's not necessarily true that everyone else will.

I would personally prefer to see some more extensive changes before landing the initial bits, but that doesn't matter too much.

brson commented:
@nikomatsakis is also thinking about how to expose compile-fail tests to users. We might want to come up with a single compile-fail scheme that works for both doc tests and unit tests.

GuillaumeGomez commented:
The first step is done, I added the `compile-fail` option for rustdoc. Now it remains to add the check of diagnostic codes. For this, I have two questions:
- Should I create a test specifically for this (if yes, how?) or do I have do complete an existing one?
- How do I run it only on diagnostic codes? How do I get them generated?

bors commented:
:umbrella: The latest upstream changes (presumably #30978) made this pull request unmergeable. Please resolve the merge conflicts.

brson commented:
@alexcrichton Again I believe it is important for Rust users to have compile-fail tests, and want to have some path forward for adding that functionality, but you sound like you are against it completely.

>     How do I run it only on diagnostic codes? How do I get them generated?

There is code somewhere for generating the html index from the error codes. I thought this was going through markdown intermediately, but apparently it is not. To run through rustdoc it should be generating markdown.

> Should I create a test specifically for this (if yes, how?) or do I have do complete an existing one?

Yes. I think what needs to happen to make this work is:

* error-index-generator should generate markdown, [not html](https://github.com/rust-lang/rust/blob/master/src/error-index-generator/main.rs#L50).
* [the build rules for error-index.html](https://github.com/rust-lang/rust/blob/master/mk/docs.mk#L220) should be updated to run the error-index markdown through rustdoc to generate the html.
* More [doc testing rules need to be added](https://github.com/rust-lang/rust/blob/master/mk/tests.mk#L830). This'll probably be a new macro like DEF_DOC_TEST that uses error-index-generator to build the markdown, then rustdoc to run the tests. You'll end up with a new set of parameterized build rules like `check-stage$(1)-T-$(2)-H-$(3)-doc-error-index-exec`.
* Add that rule to `check-stage$(1)-docs` [here](https://github.com/rust-lang/rust/blob/master/mk/tests.mk#L1014).

alexcrichton commented:
@brson sorry, but to clarify, when I said this:

> I agree, though, that testing the error index is something we should definitely do, so I'd be happy with ensuring that this is feature gated.

I meant that I'm fine moving forward with this so long as it is feature gated or somehow requires the nightly compiler (e.g. isn't present on stable Rust)

brson commented:
OK, let's move forward on this but make the feature unstable. @alexcrichton is a command-line flag sufficient, like `--allow-compile-fail-tests`, or should we try something more aggressive?

@GuillaumeGomez do the rest of the steps for testing the index make sense?

alexcrichton commented:
It may end up being an equivalent amount of code to detect something like `#![feature(rustdoc_cfail)]` in rustdoc itself, which I think would be my preferred course of action. That at least meshes with our story elsewhere and actually gates stable/nightly

GuillaumeGomez commented:
@brson: I didn't test yet, very busy with gtk-rs. I'll have time this week-end.

bors commented:
:umbrella: The latest upstream changes (presumably #31089) made this pull request unmergeable. Please resolve the merge conflicts.

brson commented:
Feature flag is ok by me.

GuillaumeGomez commented:
@brson: I added the possibility to generate markdown and updated the code when needed. Now I'm trying to figure out this "make doc" thing and why it doesn't work (something about circular references).

brson commented:
Actually, I'm not sure how the feature attribute can work since rustdoc needs to run over plain markdown files without a crate.

bors commented:
:umbrella: The latest upstream changes (presumably #31279) made this pull request unmergeable. Please resolve the merge conflicts.

GuillaumeGomez commented:
Ok, we're now super close to the end. I just have to add the last rule in the Makefile and this will be done.

GuillaumeGomez commented:
I might need some more help on the Makefile. The rules and how everything is working together is quite tricky to get.

brson commented:
I'll take a stab at the makefile integration.

brson commented:
@GuillaumeGomez for some reason I can't send you a PR (you're not in my fork list), but [here's](https://github.com/brson/rust/tree/docerr) my fixes to the makefile. I haven't been able to fully test yet because the build is so slow and it's late, but I'll let it run overnight.

We still need a way to gate this functionality so people don't use it out of tree. Since using crate stability attributes won't work (because we're parsing plain markdown files), I suggest adding an `--enable-unstable-compile-fail` flag to rustdoc. Other alternatives might be to use an environment variable that the build system uses (If we wanted to really lock it down we could use the bootstrap key). @alexcrichton to gate rustdoc compile-fail tests we can't use crate stability attributes. Other than that do you prefer a command line argument or environment variable?

GuillaumeGomez commented:
Updated.

GuillaumeGomez commented:
Updated.

brson commented:
@bors r+ Let's see how this goes...

bors commented:
:pushpin: Commit 03dd31c has been approved by `brson`

<!-- @bors r=brson 03dd31c884d7648fd6d21167779b941635af7c8c -->

alexcrichton commented on src/librustdoc/html/markdown.rs:
```diff
@@ -438,6 +440,18 @@ pub fn find_testable_code(doc: &str, tests: &mut ::test::Collector) {
     }
 }
 
+fn get_unstable_features_setting() -> bool {
```
Shouldn't this just call the function that's already in the compiler?

GuillaumeGomez replied:
This one is a bit shorter but it could (should?).

alexcrichton replied:
I have a feeling it will be much shorter if you call that one instead! It's also best to not reproduce logic wherever possible.

GuillaumeGomez replied:
Can you r- please?

alexcrichton commented on src/librustdoc/test.rs:
```diff
@@ -241,19 +242,41 @@ fn runtest(test: &str, cratename: &str, cfgs: Vec<String>, libs: SearchPaths,
+        let b_cfg = AssertRecoverSafe::new(cfg.clone());
+        let b_input = AssertRecoverSafe::new(&input);
+        let b_out = AssertRecoverSafe::new(&out);
+        let b_control = AssertRecoverSafe::new(&control);
```
Was `AssertRecoverSafe` needed for all of these variables? It may be best to match what the compiler does and spawn a new thread (or use a `monitor` function) rather than try to catch the panic here.

I doubt that much of this is *actually* recover safe

GuillaumeGomez replied:
I tried out but all variables need to be wrapped it seems. And it does recover on panic.

alexcrichton replied:
The point of the `AssertRecoverSafe` wrapper is to signal that this *is* recover-safe. I highly doubt any of these types are (they're not meant to be used in that situation), so this should follow the same strategy as the rest of the compiler (spawning a new function)

alexcrichton commented on src/librustdoc/test.rs:
```diff
@@ -241,19 +242,41 @@ fn runtest(test: &str, cratename: &str, cfgs: Vec<String>, libs: SearchPaths,
+        let b_control = AssertRecoverSafe::new(&control);
+
+        panic::recover(|| {
+            AssertRecoverSafe::new(driver::compile_input(&b_sess, &b_cstore, (*b_cfg).clone(),
```
The `AssertRecoverSafe` wrapper shouldn't be needed for the return type

GuillaumeGomez replied:
Last time I checked it did. Might worth it to retry.

alexcrichton commented:
@bors: r-

[Some last few details to attend to](https://github.com/rust-lang/rust/pull/30726#discussion_r52683510)

GuillaumeGomez commented:
@alexcrichton: Updated, you were absolutely right, `AssertRecover` wasn't needed.

GuillaumeGomez commented on src/librustdoc/html/markdown.rs:
```diff
@@ -462,6 +467,11 @@ impl LangString {
         let mut seen_rust_tags = false;
         let mut seen_other_tags = false;
         let mut data = LangString::all_false();
+        let allow_compile_fail = if let UnstableFeatures::Cheat = get_unstable_features_setting() {
```
@brson: This seems to not work. For what I understood, if it returned `UnstableFeatures::Cheat`, it means that unstable features were allowed. Did I understood wrong?

GuillaumeGomez replied:
`env::var("RUSTC_BOOTSTRAP_KEY").ok()` always returns `None`.

GuillaumeGomez commented:
Test is good! \o/

brson commented:
@bors r+

bors commented:
:pushpin: Commit eb7664b has been approved by `brson`

<!-- @bors r=brson eb7664b445d4c2a44e31b6557931dda88e7dd071 -->

bors commented:
:hourglass: Testing commit eb7664b with merge ce4b75f...

alexcrichton commented:
our linux-cross-opt builder disappeared, but otherwise all test passed, so merging manually.

michaelsproul commented:
Well done @GuillaumeGomez, this looks great!

GuillaumeGomez commented:
@michaelsproul: Thanks! :)
